sycl/unittests/helpers/PiMock.hpp

```diff
@@ -0,0 +1,197 @@
+///        function table entry
+#define _PI_API(api)                                                           \
+  template <detail::PiApiKind PiApiOffset>                                     \
+  void setFuncPtr(RT::PiPlugin *MPlugin, decltype(&::api) FuncPtr);            \
+  template <>                                                                  \
+  void setFuncPtr<detail::PiApiKind::api>(RT::PiPlugin * MPlugin,              \
+                                          decltype(&::api) FuncPtr) {          \
+    MPlugin->PiFunctionTable.api = FuncPtr;                                    \
+  }
+#include <CL/sycl/detail/pi.def>
+#undef _PI_API
+
+/// The PiMock class wraps an instance of a SYCL platform class,
+/// and manages all mock redefinitions for the underlying plugin.
+///
+/// Mock platform instances must not share the plugin resources with
+/// any other SYCL platform within the given context. Otherwise, mock
+/// redefinitions would also affect other platforms' behavior.
+/// Therefore, any plugin-related information is fully copied whenever
+/// a user-passed SYCL object instance is being mocked.
+/// The underlying SYCL platform must be a non-host plaftorm to facilitate
+/// plugin usage.
+///
+/// Simple usage examples would look like this:
+/// ```
+/// pi_result redefinePiProgramRetain(pi_program program) { /*code*/ }
+/// /*...*/
+/// unittest::PiMock Mock;
+/// Mock.redefine<PiApiKind::piProgramRetain>(redefinePiProgramRetain);
+/// platform &MockP = Mock.getPlatform();
+/// /*...*/
+/// ```
+/// ```
+/// pi_result redefinePiProgramRetain(pi_program program) { /*code*/ }
+/// /*...*/
+/// queue Q;
+/// unittest::PiMock Mock(Q);
+/// Mock.redefine<PiApiKind::piProgramRetain>(redefinePiProgramRetain);
+/// Q.submit(/* expecting mock behavior */);
+/// /*...*/
+/// ```
+class PiMock {
+public:
+  /// Default-constructs PiMock.
+  ///
+  /// Delegates to the device_selector-based constructor by passing
+  /// a default_constructor and assuming a non-host device will be
+  /// available.
+  PiMock() : PiMock(cl::sycl::default_selector{}) {}
+
+  /// Constructs PiMock from a queue.
+  ///
+  /// \param Queue is a reference to a SYCL queue to which
+  ///        the mock redefinitions will apply.
+  explicit PiMock(cl::sycl::queue &Queue)
+      : PiMock(Queue.get_device().get_platform()) {}
+
+  /// Constructs PiMock from a reference to a SYCL platform instance.
+  ///
+  /// A new plugin will be stored into the platform instance, which
+  /// will no longer share the plugin with other platform instances
+  /// within the given context. A separate platform instance will be
+  /// held by the PiMock instance.
+  ///
+  /// \param OriginalPlatform is a reference to a SYCL platform.
+  explicit PiMock(const cl::sycl::platform &OriginalPlatform) {
+    assert(!OriginalPlatform.is_host() && "PI mock isn't supported for host");
+    // Extract impl and plugin handles
+    auto ImplPtr = detail::getSyclObjImpl(OriginalPlatform);
```

Comment:
Please, do not abuse auto. Please, apply to the whole patch.

Reply:
Done. Initially, I was going to rework the test as well, but then came to realize that `getPiPlugin()` makes the type sufficiently clear.